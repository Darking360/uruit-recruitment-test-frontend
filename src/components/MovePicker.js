import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Avatar, AvatarPickerContainer } from './AvatarPicker';

let mojs = {}

// Validation in order to work with Node environments build phase
if (typeof window !== `undefined` && process.env.NODE_ENV !== 'test') {
  mojs = require('mo-js');
}

const movements = {
    '1': 'paper',
    '2': 'scissors',
    '3': 'rock'
};

const Container = styled(AvatarPickerContainer)`
    flex-direction: column;
    font-size: 1.5em;
    em, h3 {
        margin-top: 1.5rem;
    }
`;

const EnhancedAvatar = styled(Avatar)`
    background-color: white;
`;

class MovePicker extends Component {

    playBurst = () => {
        const color = '#e53935';
        if (this.moAnimation) this.moAnimation.replay();
        else {
            this.moAnimation = new mojs.Burst({
                parent:   ReactDOM.findDOMNode(this.avatar),
                radius:   { 80: 110 },
                angle:    30,
                count:    14,
                children: {
                radius:       8,
                fill:         color,
                scale:        { 1: 0, easing: 'sin.in' },
                pathScale:    [ .9, null ],
                degreeShift:  [ 13, null ],
                duration:     [ 500, 700 ],
                }
            });
            this.moAnimation.play();
        }
    }

    handleChange = () => {
        const { name, onChange, value } = this.props;
        if ((value + 1) > 3) {
            onChange(name, 1);
            return;
        }
        onChange(name, value + 1);
    }

    // Add shake phone function

    render() {
        const { value } = this.props;
        return (
            <Container ref={avatar => this.avatar = avatar}>
                <EnhancedAvatar 
                    src={`../images/${movements[value]}.png`}
                    alt='Avatar preview'
                    onClick={this.handleChange}
                />
                <em>Tap the image or shake your phone to change your attack</em>
                <h3>Attacking with {movements[value]}</h3>
            </Container>
        );
    }
}

MovePicker.propTypes = {
    value: PropTypes.number.isRequired,
    onChange: PropTypes.func.isRequired,
    name: PropTypes.string.isRequired
};

MovePicker.defaultProps = {
    value: 1,
    name: 'player1'
};

export default MovePicker;