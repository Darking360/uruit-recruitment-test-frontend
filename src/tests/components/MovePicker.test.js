import React from 'react';
import { shallow, mount } from 'enzyme';
import MovePicker from '../../components/MovePicker';

const movements = {
    '1': 'paper',
    '2': 'scissors',
    '3': 'rock'
};

const mockOnChange = jest.fn();

const mockMove = {
    name: 'player1',
    onChange: mockOnChange,
    value: 1
}

describe('Movement Picker', () => {

    it('renders without crashing', () => {
        shallow(<MovePicker {...mockMove} />);
    });

    it('matches snapshot', () => {
      const wrapper = shallow(<MovePicker {...mockMove} />);
      expect(wrapper).toMatchSnapshot();
    });

    it('call onChange with correct params', () => {
        const wrapper = mount(<MovePicker {...mockMove} />);
        wrapper.find('img').simulate('click');
        expect(mockOnChange).toHaveBeenCalled();
        expect(mockOnChange).toHaveBeenCalledWith(mockMove.name, mockMove.value + 1);
    });

    it('updates correctly its props after 1 click', () => {
        const wrapper = mount(<MovePicker {...mockMove} />);
        expect(wrapper.find('img').prop('src')).toMatch(new RegExp(`${movements[1]}`));
        wrapper.find('img').simulate('click');
        expect(mockOnChange).toHaveBeenCalledTimes(2);
        // Update props
        wrapper.setProps({ ...mockMove, value: 2 });
        expect(wrapper.find('img').prop('src')).toMatch(new RegExp(`${movements[2]}`));
    });

    it('updates correctly its props after 2 clicks', () => {
        const wrapper = mount(<MovePicker {...mockMove} />);
        expect(wrapper.find('img').prop('src')).toMatch(new RegExp(`${movements[1]}`));
        wrapper.find('img').simulate('click');
        expect(mockOnChange).toHaveBeenCalledTimes(3);
        // Update props
        wrapper.setProps({ ...mockMove, value: 3 });
        expect(wrapper.find('img').prop('src')).toMatch(new RegExp(`${movements[3]}`));
    });

});