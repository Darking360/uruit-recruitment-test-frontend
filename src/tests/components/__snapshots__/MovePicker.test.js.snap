// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movement Picker matches snapshot 1`] = `
<Styled(styled.div)>
  <Styled(styled.img)
    alt="Avatar preview"
    onClick={[Function]}
    src="../images/paper.png"
  />
  <em>
    Tap the image to change your attack
  </em>
  <h3>
    Attacking with 
    paper
  </h3>
</Styled(styled.div)>
`;
